collection-scripts/gather_monitoring

```diff
@@ -1,29 +1,208 @@
 #!/bin/bash
 
+# Gather Monitoring Metrics and Dashboards.
+# - Discovery and save Grafana dashboards, extracting the metrics' datapoints from Prometheus.
+# - Gather custom metrics (when defined on Config variables).
+# - Discovery metrics from prefix name.
+#
+# Config variables:
+# - GATHER_MONIT_START_DATE: starting human readable date string (date -d). Default: "7 days ago".
+# - GATHER_MONIT_END_DATE: ending human readable date string (date -d). Default: "now".
+# - GATHER_MONIT_QUERY_STEP: metric resolution to get from Prometheus. Default: "1m".
+# - GATHER_MONIT_CUSTOM_METRICS: list of additional metrics name to be collected. Default: Undefined
+# - GATHER_MONIT_DISCOVERY_PREFIXES: discovery metrics by prefixes. Default: ''
+# - GATHER_MONIT_TOKEN: optional service account token to gather Prometheus and Grafana API. Default: Undefined
+#
+# To create the ConfigMap:
+# $ echo -e 'GATHER_MONIT_START_DATE="15 days ago"\nGATHER_MONIT_QUERY_STEP="5m"' > env
+# $ echo -e 'GATHER_MONIT_CUSTOM_METRICS="up"' > env
+# $ echo -e 'GATHER_MONIT_DISCOVERY_PREFIXES="apiserver_"' > env
+# $ oc create configmap must-gather-env -n openshift-monitoring --from-file=env=env
+#
+# References:
+# - Prometheus API: https://prometheus.io/docs/prometheus/latest/querying/api/
+
 #Safeguards
+set -o pipefail
 set -o nounset
 set -o errexit
-set -o pipefail
 
-BASE_COLLECTION_PATH="../must-gather"
-MONITORING_PATH="${BASE_COLLECTION_PATH}/monitoring"
+function err_report {
+    echo "ERROR: on line $1 from ${0}"
+}
+trap 'err_report $LINENO' ERR
+
+declare -r GATHER_MONIT_START_DATE_DEFAULT="7 days ago"
+declare -r GATHER_MONIT_END_DATE_DEFAULT="now"
+declare -r GATHER_MONIT_QUERY_STEP_DEFAULT="1m"
```

Comment:
Should this be set higher? to say 10 min to align with the way SAR breaks down data? 
How much data will this pull back (7 days worth of data in 1 min intervals) what does that translate to in MB or size on disk?

Reply:
@sferich888  is not so easy to estimate those values. It will depends of the carnality for each metric - how often the job read those metrics on the targets (interval), how many targets the job has, how many labels/carnality each metric has, etc. 

AFIK the Prometheus api does not allow more than 11k datapoints for each query, so depending of the metric if use "7 days" && "1m" precision it will be big for one etcd metric, for instance.

Looking only to the number of labels, check the difference of time/data between collecting 1d for `up` and `etcd_request_duration_seconds_bucket` metrics.

Have a look into three examples bellow using:
[1] precision=`1m`, period=`1 day` for `up` metric
~~~
[must-gather-zqb7g] POD 2021-06-21T19:38:17.906651437Z INFO: Collecting custom metrics...
[must-gather-zqb7g] POD 2021-06-21T19:38:17.906651437Z INFO: Collecting metric /query_range?query=up
[must-gather-zqb7g] POD 2021-06-21T19:38:18.950175375Z INFO: Must-gather monitoring finished

$ du -sh mgm-1d-1m-up/*/monitoring/prometheus/metric-up.json.gz 
60K	mgm-1d-1m-up/quay-io-rhn-support-mrbraga-must-gather-sha256-7af103a5e6a1c74549dd8440ee3e611008a0504e3ade74496d6b3f7c4e7da0d4/monitoring/prometheus/metric-up.json.gz

$ zcat mgm-1d-1m-up/*/monitoring/prometheus/metric-up.json.gz  |jq -r ' .data.result[].metric.__name__'|wc -l
145
~~~
[2] precision=`1m`, period=`1 day` for `up` metric
~~~
[must-gather-r472v] POD 2021-06-21T20:01:07.337430185Z INFO: Collecting custom metrics...
[must-gather-r472v] POD 2021-06-21T20:01:07.337634071Z INFO: Collecting metric /query_range?query=up
[must-gather-r472v] POD 2021-06-21T20:01:27.791372093Z INFO: Must-gather monitoring finished


$ du -sh mgm-1d-10m-up/*/monitoring/prometheus/metric-up.json.gz 
8.0K	mgm-1d-10m-up/*/monitoring/prometheus/metric-up.json.gz

$ zcat mgm-1d-10m-up/*/monitoring/prometheus/metric-up.json.gz |jq -r ' .data.result[].metric.__name__'|wc -l
138
~~~
[3] precision=`1m`, period=`30 minutes` for `etcd_request_duration_seconds_bucket`
~~~
[must-gather-wbqs4] POD 2021-06-21T19:47:10.297731294Z INFO: Collecting metric /query_range?query=etcd_request_duration_seconds_bucket
[must-gather-wbqs4] POD 2021-06-21T19:47:15.665498273Z INFO: Must-gather monitoring finished

$ du -sh mgm-30m-1m-etcd/*/monitoring/prometheus/metric-etcd_request_duration_seconds_bucket.json.gz
2.2M	mgm-30m-1m-etcd/*/monitoring/prometheus/metric-etcd_request_duration_seconds_bucket.json.gz

$ zcat mgm-30m-1m-etcd/*/monitoring/prometheus/metric-etcd_request_duration_seconds_bucket.json.gz |jq -r ' .data.result[].metric.__name__'|wc -l
28208
~~~

Reply:
So, for that reason I believe that `GATHER_MONIT_START_DATE_DEFAULT` is already high, I would love to leave it by default `3 days` and if needed it can be chanced by CM.

Reply:
Also for that reason I cut a lot of "features" from initial PR that could impact in performance, like "Grafana metrics discovery". It could be a problem if enabled by default. IMO We need to improve it before become a option to run in CU envs.